### PR TITLE
Log exports of search results to the database

### PR DESCRIPTION
**Pull request by reupen**
### Description of change

This adds a new model UserEvent (and a new app user_event_log) that can be used to track certain events that happen when a user is using the system.

This is used to log when search results are exported to a CSV file by a user.

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.rst](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.rst) for instructions
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [x] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [x] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/search/views.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
 
 from datahub.core.csv import create_csv_response
 from datahub.oauth.scopes import Scope
+from datahub.user_event_log.contants import EVENTS
```
Comment (elcct) [inline datahub/search/views.py]: `constants` ?

File: datahub/user_event_log/utils.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+from datahub.user_event_log.models import UserEvent
+
+
+def record_user_event(request, event, model=None, data=None):
+    """
+    Records a user event in the database.
+
+    UserEvent.data is configured to use the default JSON serialiser which means data should be
+    JSON-serialisable by json.dumps (so datetimes and UUIDs are not allowed).
```
Comment (elcct) [inline datahub/user_event_log/utils.py]: Is this true? It seems that the tests pass UUIDs and datetimes and the data field uses `DjangoJSONEncoder` as encoder.

Comment (reupen) [reply datahub/user_event_log/utils.py]: Thanks, it was true before I changed the encoder :slightly_smiling_face:

File: datahub/search/views.py (hunk 1 above)
Comment (reupen) [reply datahub/search/views.py]: Whoops, thanks!

File: datahub/user_event_log/models.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    User event.
+
+    Used to keep a record of specific events that have occurred while the user was using the
+    system (e.g. the user exported data from search).
+
+    Not intended as a replacement for logging, but for cases where we need to record data in a
+    more structured fashion and retain it for a longer period of time.
+    """
+
+    id = models.BigAutoField(primary_key=True)
```
[the remark this answers is not in the document]
Comment (reupen) [reply datahub/user_event_log/models.py]: Good question. We have other tables without UUIDs and it seemed more consistent with the other log tables we have (reversion_*, django_admin_log). Could change it if we think there's a benefit to using UUID here.

[the remark this answers is not in the document]
Comment (elcct) [reply datahub/user_event_log/models.py]: For me, the benefit would be consistency, in the sense that reversion_*, django_admin_log are 3rd party, so I am not sure keeping consistency with them over Leeloo is worth doing.

[the remark this answers is not in the document]
Comment (reupen) [reply datahub/user_event_log/models.py]: No so fussed about it, but InvestmentProjectStageLog does not use UUID either.

[the remark this answers is not in the document]
Comment (elcct) [reply datahub/user_event_log/models.py]: I thought UUID for a model such strongly tied to Investment Project would be an overkill. I am okay if we stay with `BigAutoField` too.

File: datahub/user_event_log/models.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        'company.Advisor',
+        on_delete=models.CASCADE,
+        related_name='user_events',
+    )
+    event = models.CharField(max_length=settings.CHAR_FIELD_MAX_LENGTH, choices=EVENTS)
+    content_type = models.ForeignKey(
+        'contenttypes.ContentType',
+        null=True,
+        on_delete=models.CASCADE,
+    )
```
[the remark this answers is not in the document]
Comment (reupen) [reply datahub/user_event_log/models.py]: It's not really needed anyway. I can just get rid of it.

**Pull request merged**